### PR TITLE
Add sqlite

### PR DESCRIPTION
**Pull request by djih**
I am using an iOS wrapper for objective c: https://github.com/ccgus/fmdb

### CODE REVIEW

File: Amplitude/AMPDatabaseHelper.m [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+        if (upToId > 0 && limit > 0) {
+            querySQL = [NSString stringWithFormat:GET_EVENT_WITH_UPTOID_AND_LIMIT, ID_FIELD, EVENT_FIELD, table, ID_FIELD, upToId, limit];
+        } else if (upToId > 0) {
+            querySQL = [NSString stringWithFormat:GET_EVENT_WITH_UPTOID, ID_FIELD, EVENT_FIELD, table, ID_FIELD, upToId];
+        } else if (limit > 0) {
+            querySQL = [NSString stringWithFormat:GET_EVENT_WITH_LIMIT, ID_FIELD, EVENT_FIELD, table, limit];
+        } else {
+            querySQL = [NSString stringWithFormat:GET_EVENT, ID_FIELD, EVENT_FIELD, table];
+        }
+        FMResultSet *s = [db executeQuery:querySQL];
```
Comment (curtisliu) [inline Amplitude/AMPDatabaseHelper.m]: avoid using single letter variable names

File: Amplitude/AMPDatabaseHelper.m [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+        if (s == nil) {
+            NSLog(@"getEvents from table %@ failed: %@", table, [db lastErrorMessage]);
+            [db close];
+            [self resetDB:NO];
+            return;
+        }
+
+        while ([s next]) {
```
Comment (curtisliu) [inline Amplitude/AMPDatabaseHelper.m]: does the result set need to be closed? (not sure how the interface works)

Comment (djih) [inline Amplitude/AMPDatabaseHelper.m]: How cool is this, closing the DB closes all open resultsets: https://github.com/ccgus/fmdb/commit/ae8622d5b05747bc4573b1f0520b1b346a94575e

File: Amplitude/Amplitude.m [hunk 3]
```diff
@@ -207,19 +217,24 @@ - (id)init
...
+
+            // update the database
+            if (oldDBVersion < kAMPDBVersion) {
+                [[AMPDatabaseHelper getDatabaseHelper] upgrade:oldDBVersion newVersion:kAMPDBVersion];
+                [_propertyList setObject:[NSNumber numberWithInt:kAMPDBVersion] forKey:DATABASE_VERSION];
+                [self savePropertyList];
             }
 
+            // migrate all of old _eventsData object to database store
+            [self migrateEventsDataToDB];
```
Comment (curtisliu) [inline Amplitude/Amplitude.m]: I think it would be safest to wrap this with a version < X check

File: Amplitude/Amplitude.m [hunk 4]
```diff
@@ -239,6 +254,44 @@ - (id)init
...
+        [dbHelper insertOrReplaceKeyValue:USER_ID value:userId];
+    }
+    NSNumber *optOut = [eventsData objectForKey:OPT_OUT];
+    if (optOut != nil) {
+        [dbHelper insertOrReplaceKeyLongValue:OPT_OUT value:optOut];
+    }
+    NSString *deviceId = [eventsData objectForKey:DEVICE_ID];
+    if (deviceId != nil) {
+        [dbHelper insertOrReplaceKeyValue:DEVICE_ID value:deviceId];
+    }
```
Comment (curtisliu) [inline Amplitude/Amplitude.m]: last event time and previous session id need migrating too

File: Amplitude/AMPDatabaseHelper.m (hunk 1 above)
Comment (djih) [inline Amplitude/AMPDatabaseHelper.m]: fixed

File: Amplitude/Amplitude.m (hunk 3 above)
Comment (djih) [inline Amplitude/Amplitude.m]: Done

File: Amplitude/Amplitude.m (hunk 4 above)
Comment (djih) [inline Amplitude/Amplitude.m]: Added

**Comment by djih** [comment]
PTAL @curtisliu

**Comment by curtisliu** [comment]
LGTM

**Pull request merged**